wrangler-service/src/main/java/io/cdap/wrangler/service/DataPrepService.java

```diff
@@ -74,4 +84,49 @@ protected void configure() {
     addHandler(new DataModelHandler());
     addHandler(new WorkspaceHandler());
   }
+
+  @Override
+  public void initialize(SystemServiceContext context) {
+    // only do the upgrade on first instance to avoid transaction conflict
+    if (context.getInstanceId() != 0) {
+      return;
+    }
+
+    UpgradeStore upgradeStore = new UpgradeStore(context);
+    WorkspaceStore wsStore = new WorkspaceStore(context);
+    boolean isConnDone = upgradeStore.isEntityUpgradeComplete(UpgradeEntityType.CONNECTION);
```

Comment:
this logic assumes one instance of the service, which is the normal setup but not necessarily the case for all setups. Does the ServiceContext expose the instance id? If so you can check that to make sure only instance 0 upgrades.

Reply:
I added the logic of checking instance id here - https://github.com/data-integrations/wrangler/pull/510/files/b1546431b6c791713cd71577d381e9a1a00dea0f..30c1ddf35ccd97a057ec1f3e09f5a42c256284de#diff-4a6d56260e39433bf54fa3ac1e682aa59ee6b1cd6218003d07a3b882cce14795R91-R93, it will only upgrade on instance 0